src/sql/src/plan/statement.rs

```diff
@@ -446,13 +460,28 @@ fn handle_show_databases(
     scx: &StatementContext,
     ShowDatabasesStatement { filter }: ShowDatabasesStatement,
 ) -> Result<Plan, anyhow::Error> {
-    let rows = scx
-        .catalog
-        .list_databases()
-        .map(|database| vec![Datum::from(database)])
-        .collect();
-
-    finish_show_where(scx, filter, rows, &SHOW_DATABASES_DESC)
+    // RelationExpr::Get must stay in sync with the definition of
+    // MZ_DATABASES in BUILTINS.
+    let expr = RelationExpr::Project {
+        input: Box::new(RelationExpr::Get {
+            id: expr::Id::Global(GlobalId::System(2011)), // Hard-coded system id for mz_databases.
+            typ: RelationType {
+                column_types: vec![
+                    ColumnType {
+                        nullable: false,
+                        scalar_type: ScalarType::String,
+                    },
+                    ColumnType {
+                        nullable: false,
+                        scalar_type: ScalarType::String,
+                    },
+                ],
+                keys: vec![],
+            },
+        }),
+        outputs: vec![1], // Only return the database name, not the id.
+    };
```

Comment:
The amount of hardcoding here of magic numbers and types is worrying to me. What if instead you synthesized a `sql::ast::Select` that specified `SELECT database FROM mz_catalog.mz_databases WHERE ...`?

There's an example of building up `Select` statements programmatically here: https://github.com/MaterializeInc/materialize/blob/be9d85fabb7a2713b6426b1482ec0901f3671d76/src/sql/src/plan/transform_ast.rs#L324-L350

You can add more builder helper methods to the AST types as well!

Reply:
I think [this commit](https://github.com/MaterializeInc/materialize/pull/4033/commits/4b69b38cad63348b1124cb2c645d58f697fe04be) does what you mean!